CDAP-2556: Add support for HBase1.1 in CDAP

rohitsinha54 opened the pull request.

poornachandra commented on cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/increment/hbase11/IncrementSummingScanner.java:
```diff
@@ -0,0 +1,368 @@
+
+  private boolean nextInternal(List<Cell> cells, ScannerContext scannerContext) throws IOException {
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("nextInternal called with limit=" + scannerContext);
```
Since the message says "with limit", no need to print the whole `scannerContext`, right?

rohitsinha54 replied:
As discussed in person. The only thing which ScannerContext exposes is toString which will print the limit with progress. So keeping this.

poornachandra commented on cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/increment/hbase11/IncrementSummingScanner.java:
```diff
@@ -0,0 +1,368 @@
+    int addedCnt = 0;
+    baseScanner.startNext();
+    Cell cell;
+    while ((cell = baseScanner.peekNextCell(getBatch())) != null && (getBatch() <= 0 || addedCnt < getBatch())) {
```
We need to use limit from `scannerContext` everwhere rather than `getBatch` when `scannerContext` is available. 

Actually, I think the newly introduced method `getBatch` is not required at all. Any reason for adding it?

poornachandra commented on cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/util/hbase/HBase11TableUtil.java:
```diff
@@ -0,0 +1,356 @@
+  @Override
+  public Class<? extends Coprocessor> getTransactionDataJanitorClassForVersion() {
+    CodeSource codeSource = HBase11TableUtil.class.getProtectionDomain().getCodeSource();
+    LOG.error("HBase11TableUtil is getting loaded from {}", codeSource == null ? "<null>" : codeSource.getLocation());
```
All these log messages need to be removed.

poornachandra commented on cdap-hbase-compat-1.1/src/test/java/co/cask/cdap/data2/increment/hbase11/IncrementHandlerTest.java:
```diff
@@ -0,0 +1,206 @@
+
+    @Override
+    public boolean flush() throws IOException {
+      HRegion.FlushResult result = region.flushcache(true, false);
```
It would be a good idea to add a comment here saying why `flushCache` is `(true, false)`.

poornachandra commented:
@rohitsinha54 Few comments, please address them.

poornachandra commented on cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/increment/hbase11/IncrementSummingScanner.java:
```diff
@@ -0,0 +1,370 @@
+    int addedCnt = 0;
+    baseScanner.startNext();
+    Cell cell;
+    while ((cell = baseScanner.peekNextCell(scannerContext)) != null && (getBatch() <= 0 || addedCnt < getBatch())) {
```
All get batch size requests should come from scanner context. In this case use internal scanner to get the batch size.

rohitsinha54 commented:
This PR needs some changes to handle the API change which affect IncrementSummingScanner according to the review comments. I will fix them today and re-open this PR.

rohitsinha54 commented:
Addressed the comments and also the issue cause by newly introduced ScannerContext.

The scan in HBase has change in 1.1 : https://blogs.apache.org/hbase/entry/scan_improvements_in_hbase_1

It introduces a ScannerContext which takes three kind of limits: batchLimit, sizeLimit and timeLimits. There is no way to get these limits from the ScannerContext as the getter methods are package private.

Now in our IncrementSummingScanner we implement RegionScanner and delegate the next calls to the RegionScanner with limits (now ScannerContext). We cannot depend on the RegionScanner limit tracking for batch size (limit on number of columns) as it does not differentiate between an increment and non-increment cells.
We need to do the limit tracking for cells to identify different columns and progress towards limit in IncrementSummingScanner which differentiate between a increment and non-increment cell and to do this we need to know the batch limit given in the ScannerContext. 

The 	"Add ScannerContextReader for HBase1.1 support" commit introduces a ScannerContextReader class in the same package as ScannerContext in HBase and delegates the call to get the limits to ScannerContext. This allow us to get the limits and use in it IncrementSummingScanner.

Also added a test which explicitly test for limit and the columns retrieved.

rohitsinha54 commented:
New build running here: http://builds.cask.co/browse/CDAP-DUT2754-5

poornachandra commented on cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/increment/hbase11/IncrementSummingScanner.java:
```diff
@@ -0,0 +1,379 @@
+    // So, we read all the cells from the internal scanner in batches of limit and do the tracking for columns here
+    // and stop when we have reached the limit.
+    ScannerContextReader scannerContextReader = new ScannerContextReader(scannerContext);
+    int limit = scannerContextReader.getBatchLimit();
```
It would be a good idea to move the above lines into a private method called `getBatchLimit()` or something. Then the six lines of comments can also move to the private method.

chtyim commented:
LGTM

rohitsinha54 commented:
Waiting for a final build here: http://builds.cask.co/browse/CDAP-DUT2754-6
Will merge as soon as build passes.
